### PR TITLE
[Serializer] Fix argument object denormalization

### LINKED ISSUE #20670: [Serializer] Wrong deserialization with object as constructor parameter
Hello,

I start with a bug before a PR because I have no time for now and it looks very weird.

In the latest RC this line looks very weird : https://github.com/symfony/symfony/blob/v3.2.0-RC2/src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php#L340

I don't get how it could work (and it definitively do not work in my app). 

We should run a `denormalize` operation instead of  `deserialize` operation. All data should be at least decoded from json, no ?

I get also ` Warning: json_decode() expects parameter 1 to be string, array given in /var/app/vendor/symfony/symfony/src/Symfony/Component/Serializer/Encoder/JsonDecode.php on line 88` that which make me say I'm right.

Thanks

edit: Replacing the deserialize by denormalize make my app working

Here is a gist to reproduce : https://gist.github.com/tyx/ed50f6e525494909249b571c76360c36

### PR DESCRIPTION
**Pull request by ogizanagi**
| Q             | A
| ------------- | ---
| Branch?       | 3.2
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #20670
| License       | MIT
| Doc PR        | N/A

Fixes #20670. I've seen https://github.com/symfony/symfony/pull/19277#discussion-diff-69389638 so I think it's the right thing to do, but I didn't follow the thread at the time, so I may have missed something.

Ping @theofidry, @dunglas.

### CODE REVIEW

**Comment by theofidry** [comment]
👍 I had this patch in mind since day one but never found the time to do it thanks very much 🎉 

I personally hesitated between:

- This solution
- Having a null encoder (the patch I used on my side meanwhile)
- Adding a `DenormalizerAware` which IMO is more appropriate than `SerializerAware`

**Comment by tyx** [comment]
Thanks for this fix, looks great for me. 

For me this feature was the most expected of the 3.2, I hope we can have this one merged within the first release !

**Comment by ogizanagi** [comment]
> hope we can have this one merged within the first release !


@tyx : That's already too late for this 😅  http://symfony.com/blog/symfony-3-2-0-released

**Comment by tyx** [comment]
😢

**Comment by dunglas** [comment]
IMO it should be merged as a bug fix in 3.1 (and so we'll be in the next minor version of 3.2).

File: src/Symfony/Component/Serializer/Tests/SerializerTest.php [hunk 1]
```diff
@@ -336,6 +336,22 @@ public function testDenormalizerAware()
 
         new Serializer(array($denormalizerAware));
     }
+
+    public function testDeserializeObjectConstructorWithObjectTypeHint()
+    {
+        $jsonData = '{"bar":{"value":"baz"}}';
+
+        $serializer = new Serializer(
+            array(
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Tests/SerializerTest.php]: Can be inlined (same below).

**Review by dunglas — APPROVED** [review]

**Comment by dunglas** [comment]
👍

**Comment by ogizanagi** [comment]
> IMO it should be merged as a bug fix in the 3.1 branch (and so we'll be in the next minor version of 3.2).

@dunglas : 3.1? But it was introduced in https://github.com/symfony/symfony/pull/19277 which has been merged in 3.2 only 😕

**Comment by dunglas** [comment]
Sorry, you're right, it should be merged in 3.2.

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php [hunk 2]
```diff
@@ -336,8 +337,11 @@ protected function instantiateObject(array &$data, $class, array &$context, \Ref
                     $parameterData = $data[$key];
                     try {
                         if (null !== $constructorParameter->getClass()) {
+                            if (!$this->serializer instanceof DenormalizerInterface) {
+                                throw new LogicException(sprintf('Cannot create an instance of %s from serialized data because the injected serializer is not a denormalizer', $constructorParameter->getClass()));
```
Comment (xabbuh) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: Should we add the FQCN of the serializer here to provide some more context information?

Comment (ogizanagi) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: Similar exceptions above do not add such information (I guess the stacktrace is enough for this)

Comment (ogizanagi) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: Hmm, but other ones are runtime exceptions. Whereas here we need to fix the code/configuration and the stacktrace will only show `AbstractNormalizer`. So maybe you're right 👍

Comment (ogizanagi) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: Done

**Comment by dunglas** [comment]
Thanks for fixing this bug @ogizanagi.

**Pull request merged**